[guangxuan] iP

guangxuan opened the pull request:
**Duke Pro**

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)
DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [ ]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)
___
If you Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

<table>
  <tr>
    <th>Col 1</th>
    <th>Col 2</th>
  </tr>
  <tr>
    <td>item</td>
    <td>item</td>
  </tr>
  <tr>
    <td>item</td>
    <td>item</td>
  </tr>
</table>

Commit:
Add Gradle support

guangxuan committed:
feat: add Level-1

guangxuan committed:
feat: add Level-2

guangxuan committed:
feat: add Level-3

guangxuan committed:
feat: add Level-4

guangxuan committed:
feat: add text ui testing

guangxuan committed:
feat: add Level-5

guangxuan committed:
feat: add Level-6

guangxuan committed:
Add TaskType enum

guangxuan committed:
Add Storage

guangxuan committed:
Add Event and Deadline DateTime

guangxuan committed:
Merge branch 'branch-Level-7'

guangxuan committed:
Merge branch 'branch-Level-8'

guangxuan committed:
Fix Storage DateTime

guangxuan committed:
Add More OOP

guangxuan committed:
Add duke package

guangxuan committed:
Add JUnit tests

guangxuan committed:
Fix FileNotFoundException

guangxuan committed:
Add build jar artifact

guangxuan committed:
Add Javadocs

guangxuan committed:
Refactor code to follow coding standard

guangxuan committed:
Add Find command

guangxuan committed:
Merge branch 'branch-level-9'

guangxuan committed:
Merge branch 'branch-A-JavaDoc'

guangxuan committed:
merge branch 'branch-A-CodingStandard'

guangxuan committed:
Merge remote-tracking branch 'origin/add-gradle-support'

guangxuan committed:
Add checkstyle

guangxuan committed:
Add JavaFX tutorial part 3

guangxuan committed:
Add JavaFX tutorial part 4

guangxuan committed:
Merge branch 'branch-Level-10'

waynetee commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,33 @@
+     * @param taskList contains the task list
+     * @param ui handles interactions with the user
+     * @param storage handles loading tasks from the file and saving tasks in the file
+     * @throws DukeException
+     * @throws Exception
```
Should the JavaDoc include the scenario in which the exception will be thrown?

guangxuan replied:
Yes.

waynetee commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,33 @@
+    private TriConsumer func;
+    private boolean isExit;
+
+    /**
+     * @param func function that has to be called with taskList, ui, storage
+     * @param isExit true iff func is an exit function
+     */
```
I think you need to add punctuation to all your javadoc as per the style guide :-)

guangxuan replied:
Yes.

waynetee commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+ * Represents a Task with an additional dateTime field storing a deadline
+ */
+public class Deadline extends Task {
+    private LocalDate dateTime;
```
does your datetime actually store the time?

guangxuan replied:
No. I will rename ```dateTime``` to ```date``` for accuracy.

waynetee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    /**
+     * Run commands obtained from GUI to obtain output
```
```suggestion
     * Runs commands obtained from GUI to obtain output
```

guangxuan replied:
Yes.

waynetee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,91 @@
+        }
+    }
+
+    /**
+     * Main loop for interactions with user
+     */
```
According to the coding standard

> In method header comments, the first sentence should start in the form Returns ..., Sends ..., Adds ... (not Return or Returnning etc.)

guangxuan replied:
Yes.

waynetee commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,128 @@
+ */
+public class Parser {
+    /**
+     * Parses a line of input to a Command object
+     * @param fullCommand string containing exactly one line of input
```
```suggestion
     * Parses a line of input to a Command object
     *
     * @param fullCommand string containing exactly one line of input
```

guangxuan replied:
Yes.

waynetee reviewed (approved):
Nice bot, nice use of functional interface too.

LGTM :)

guangxuan committed:
Fix coding standard violations

Codebase violates coding standards.

Coding standards improves readability of code.

Let's
* Add punctuation to javadocs
* Add appropriate endlines to javadocs
* Change javadocs description to start with verbs

guangxuan committed:
Add Assertions

Assertions can document important assumptions that should hold at various points in the code

Let's
* Add assertions to Duke, MainWindow, Ui Class

guangxuan committed:
Add C-Undo

Current solution does not easily allow fixing of past user mistakes.

The ability to undo and redo will allow users ease of correcting mistakes.

Let's
* Add HistoryManager to handle historical logic
* Add HistoryManager when running Commands

dillontkw commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,58 @@
+        return scanner.nextLine();
+    }
+
+    public void showln(String s) {
```
Perhaps naming this method as showLine would be better? I think it would be more consistent with the other method names in this class

guangxuan replied:
Yes.

dillontkw commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,58 @@
+
+    /**
+     * Prints stack trace of other error messages
+     * @param e Erorr thrown by Java
```
```suggestion
     * @param e Error thrown by Java
```

guangxuan replied:
Yes.

dillontkw reviewed (commented):
Hi Guang Xuan,

I like your usage of functional interfaces in your project. Besides the points that wayne has mentioned, I didn't notice any violations of other code quality guidelines. Great job! 🙂

guangxuan committed:
Add Dillion PR comments

guangxuan committed:
Merge pull request #1 from guangxuan/branch-A-Assertions

Add Assertions

guangxuan committed:
Merge pull request #2 from guangxuan/branch-A-CodeQuality

Branch a code quality

guangxuan committed:
Optimize import orders

guangxuan committed:
A-BetterGui

guangxuan committed:
A-UserGuide

guangxuan committed:
Fix UI Exception bug and dialog height bug

guangxuan committed:
Rename Ui.png

guangxuan committed:
Fix user guide

guangxuan committed:
fix javaFX on linux and mac
